internal/datastore/postgres/strictreader.go

```diff
@@ -0,0 +1,65 @@
+package postgres
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/jackc/pgx/v5"
+	"github.com/jackc/pgx/v5/pgconn"
+
+	"github.com/authzed/spicedb/internal/datastore/common"
+	pgxcommon "github.com/authzed/spicedb/internal/datastore/postgres/common"
+)
+
+const pgInvalidArgument = "22023"
+
+// strictReaderQueryFuncs wraps a DBFuncQuerier and adds a strict read assertion to all queries.
+// This assertion ensures that the transaction is not reading from the future or from a
+// transaction that has not been committed on the replica.
+type strictReaderQueryFuncs struct {
+	wrapped  pgxcommon.DBFuncQuerier
+	revision postgresRevision
+}
+
+func (srqf strictReaderQueryFuncs) ExecFunc(ctx context.Context, tagFunc func(ctx context.Context, tag pgconn.CommandTag, err error) error, sql string, args ...any) error {
+	// NOTE: it is *required* for the pgx.QueryExecModeSimpleProtocol to be added as pgx will otherwise wrap
+	// the query as a prepared statement, which does *not* support running more than a single statement at a time.
+	return srqf.rewriteError(srqf.wrapped.ExecFunc(ctx, tagFunc, srqf.addAssertToSQL(sql), append([]interface{}{pgx.QueryExecModeSimpleProtocol}, args...)...))
```

Comment:
I have concerns about us forcing`pgx.QueryExecModeSimpleProtocol`, but I understand it would be a tradeoff that has to be made to make replicas work.

For context, we already force this in `QueryRelationships` [after this investigation](https://medium.com/engineering-at-birdie/puzzling-postgres-a-story-of-solving-an-unreproducible-performance-issue-778075ed7998), which resulted in https://github.com/authzed/spicedb/pull/496, but this would expand this behaviour to all datastore methods. I highly doubt it, but makes me wonder if the perf degradation would exceed the "always check revision" strategy used in MySQL.

Reply:
I imagine it might be a bit slower, but nothing substantial

Reply:
only one way to find out I guess